onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder.cc

```diff
@@ -2694,6 +2694,108 @@
+    model_builder.AddInitializerToSkip(inputs[2].node_arg.Name());  // constant_value
+  }
+}
+
+Status PadOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder, const NodeUnit& node_unit) const {
+  auto& shaper = model_builder.GetShaper();
+  const auto& operand_indices = model_builder.GetOperandIndices();
+  const auto& operand_types = model_builder.GetOperandTypes();
+  const auto& inputs = node_unit.Inputs();
+  const auto& outputs = node_unit.Outputs();
+
+  std::vector<uint32_t> input_indices{};
+
+  // `data` input
+  const auto& data = inputs[0].node_arg.Name();
+  input_indices.push_back(operand_indices.at(data));
+
+  // `pads` input
+  // convert from [begin_1, begin_2, ..., end_1, end_2, ...] to [begin_1, end_1, begin_2, end_2, ...]
+  // convert from int64_t to int32_t
+  const auto& data_shape = shaper[data];
+  const uint32_t data_rank = SafeInt<uint32_t>(data_shape.size());
+
+  const auto& pads = inputs[1].node_arg.Name();
+  const auto* pads_initializer = model_builder.GetConstantInitializer(pads);
+  ORT_RETURN_IF_NOT(pads_initializer, "pads must be a constant");
+  ORT_ENFORCE(pads_initializer->data_type() == ONNX_NAMESPACE::TensorProto_DataType_INT64);
+
+  std::vector<uint8_t> pads_initializer_raw_data{};
+  ORT_RETURN_IF_ERROR(utils::UnpackInitializerData(*pads_initializer, pads_initializer_raw_data));
+  ORT_RETURN_IF_NOT(pads_initializer_raw_data.size() == 2 * data_rank * sizeof(int64_t),
+                    "Expected pads initializer size in bytes: ", 2 * data_rank * sizeof(int64_t),
+                    ", actual: ", pads_initializer_raw_data.size());
+
+  std::vector<int32_t> converted_pads_data{};
+  converted_pads_data.reserve(2 * data_rank);
+
+  auto copy_and_convert = [](const void* raw_i64_src,
+                             std::back_insert_iterator<decltype(converted_pads_data)> i32_dst) {
+    int64_t i64;
+    memcpy(&i64, raw_i64_src, sizeof(i64));
+    *i32_dst = SafeInt<int32_t>(i64);
+  };
+
+  for (size_t i = 0; i < data_rank; ++i) {
+    copy_and_convert(&pads_initializer_raw_data[i * sizeof(int64_t)],
+                     std::back_inserter(converted_pads_data));
+
+    copy_and_convert(&pads_initializer_raw_data[(i + data_rank) * sizeof(int64_t)],
+                     std::back_inserter(converted_pads_data));
+  }
+
+  const Shape converted_pads_shape{data_rank, 2};
+  const OperandType converted_pads_operand_type{Type::TENSOR_INT32, converted_pads_shape};
+  ORT_RETURN_IF_ERROR(model_builder.AddOperandFromPersistMemoryBuffer(pads, converted_pads_data.data(),
+                                                                      converted_pads_operand_type));
+  input_indices.push_back(operand_indices.at(pads));
+
+  // `constant_value` input
+  float pad_value = 0.0f;
+  if (inputs.size() > 2 && inputs[2].node_arg.Exists()) {
+    const auto constant_value = inputs[2].node_arg.Name();
+    const auto* constant_value_initializer = model_builder.GetConstantInitializer(constant_value);
```

Comment:
const auto&?

Reply:
GetConstantInitializer() returns a possibly null pointer. currently, in op support checker it's actually checking that the (possibly not constant) initializer exists.
once we update op support checker to check for constant initializers I think we can directly dereference the pointer here.

Reply:
Sorry - codeflow moved the line. I meant for the constant_value that is using Name() which I think returns a std::string&.